src/components/utils/src/timer.cc

```diff
@@ -44,175 +44,180 @@ CREATE_LOGGERPTR_GLOBAL(logger_, "Utils")
-  DCHECK_OR_RETURN_VOID(timer_);
-  is_running_ = true;
+timer::Timer::TimerDelegate::TimerDelegate(Timer* timer)
+    : timer_(timer)
+    , timeout_milliseconds_(0)
+    , state_lock_(true)
+    , stop_flag_(false)
+    , restart_flag_(false)
+    , is_started_flag_(false)
+    , is_repeatable_(false) {}
+
+timer::Timer::TimerDelegate::~TimerDelegate() {
+  timer_ = NULL;
 }
 
-bool timer::Timer::StopUnsafe() {
-  LOG4CXX_DEBUG(logger_, "Stopping timer  " << name_);
-  //  Destroing of posix timer
-  if (StopPosixTimer(timer_)) {
-    is_running_ = false;
-    return true;
+void timer::Timer::TimerDelegate::threadMain() {
+  using sync_primitives::ConditionalVariable;
+  sync_primitives::AutoLock auto_lock(state_lock_);
+  stop_flag_ = false;
+  // Notify that we've already started
+  TimerDelegate::is_started_flag_ = true;
+  TimerDelegate::starting_condition_.NotifyOne();
+
+  while (!stop_flag_) {
+    // Sleep
+    int32_t wait_milliseconds_left = TimerDelegate::get_timeout();
+    LOG4CXX_DEBUG(logger_,
+                  "Milliseconds left to wait: " << wait_milliseconds_left);
+    ConditionalVariable::WaitStatus wait_status =
+        termination_condition_.WaitFor(auto_lock, wait_milliseconds_left);
+    // Quit sleeping or continue sleeping in case of spurious wake up
+    if (ConditionalVariable::kTimeout == wait_status ||
+        wait_milliseconds_left <= 0) {
+      LOG4CXX_DEBUG(logger_,
+                    "Timer has finished counting. Timeout(ms): "
+                        << wait_milliseconds_left);
+      DCHECK(timer_);
+      timer_->OnTimeout();
+    } else {
+      LOG4CXX_DEBUG(
+          logger_,
+          "Timeout reset force (ms): " << TimerDelegate::timeout_milliseconds_);
+    }
+
+    if (is_repeatable_) {
+      // it's repeatable just looping here
+      continue;
+    }
+    if (restart_flag_) {
+      // It's one-shoot timer, it've updated timeout
+      // need one more shoot
+      restart_flag_ = false;
+      continue;
+    }
+    // One-shoot timer wihout update timeout, just exit
+    return;
   }
-  return false;
 }
 
-timer::Milliseconds timer::Timer::GetTimeout() const {
-  sync_primitives::AutoLock auto_lock(lock_);
-  return timeout_ms_;
+void timer::Timer::TimerDelegate::exitThreadMain() {
+  LOG4CXX_DEBUG(logger_, "Exit from timer thread.");
+  ShouldBeStopped();
+  termination_condition_.NotifyOne();
+}
+
+void timer::Timer::TimerDelegate::SetTimeOut(
+    const Milliseconds timeout_milliseconds) {
+  timeout_milliseconds_ = timeout_milliseconds;
+  termination_condition_.NotifyOne();
+}
+
+void timer::Timer::TimerDelegate::ShouldBeStopped() {
+  stop_flag_ = true;
+  restart_flag_ = false;
+}
+
+void timer::Timer::TimerDelegate::ShouldBeRestarted() {
+  restart_flag_ = true;
+}
+
+bool timer::Timer::TimerDelegate::IsGoingToStop() const {
+  return stop_flag_;
+}
+
+void timer::Timer::TimerDelegate::WaitUntilStart() {
+  sync_primitives::AutoLock auto_lock(TimerDelegate::state_lock_);
```

Comment:
@AGritsevich, why the lock is out of loop? This way you protect the atomic flag. Can you move lock inside the loop?